### ISSUE TITLE
Add assignment viewed event

### ISSUE THREAD

**Issue by micaherne** [davidpesce/moodle-logstore_xapi#46]
**Motive**
"Module viewed" events are raised when a student views any module except for mod_assign. A student viewing an assignment is as, or more, important to know about than viewing any other module, so it would be a useful addition.

**Result**
A statement sent to the LRS when a student clicks on an assignment in Moodle and goes to the assignment view page.

**Additional information**
I think this may have been missed out as there isn't a course_module_viewed event as such, unlike the other modules, but there are other events (e.g. \mod_assign\event\submission_status_viewed) that could potentially be used in place of it.

**Comment by jlowe64** [comment]
Looking into the events available, and the current status of the mod_assign plugin, there doesn't appear to be any viewed events in moodle at all. It may need to be an issue brought up with Moodle HQ and integrated into Moodle core. 

Unless the list I've looked at is completely out of date. Please correct me.

**Comment by garemoko** [comment]
This brings up an interesting challenge- how we address Moodle events that are simply missing. I guess it's out of scope of this plugin, but it seems like a problem that's likely to come up again and again as we look to extend what we're tracking.

**Comment by jlowe64** [comment]
CLC has just started adding them to core. We can't wait on Moodle to act on it. We are currently adding scorm events that will be in 3.1.

**Comment by garemoko** [comment]
@jlowe64 nice!

**Comment by joshwillcock** [comment]
I have looked into adding assignment viewed for the Mod assign. However the only part which may be viewed which is relevant to us would be viewing the assignment overview which triggers: "mod_assign\event\submission_status_viewed" as soon as you land on the assign/view.php page which has the description, deadlines and ability to upload. @micaherne was this what you wanted to track or was there something else specifically missing?

**Comment by jlowe64** [comment]
@joshwillcock thanks for looking into that. @micaherne can you confirm the exact bit of information you'd like to track there if it's anything beyond a simple view?

**Comment by micaherne** [comment]
@joshwillcock @jlowe64 yes it was just a simple view that needs to be tracked, and I totally agree that submission_status_viewed seems to be the Moodle event to use.

It's interesting that mod_assign doesn't have a generic course_module_viewed event, although I can kind of see why it doesn't since most of the view actions throw their own more specific event.

I do think the logstore needs to send _something_ to the LRS when the student views the activity, given that all the other activities do, and mod_assign is arguably the most important of them all!

**Comment by micaherne** [comment]
I've raised an issue in the Moodle tracker about the generic event: https://tracker.moodle.org/browse/MDL-53035.

It might be that there are some rules I'm not aware of about not raising multiple events for a single view or something, but if nobody objects I'll see about getting a pull request sorted out there.

**Comment by jlowe64** [comment]
@micaherne I've added this event to the list. I think it's sufficient! They seem to be very particular about the terms they use for events so it may end up an unnecessary argument on their end.

**Comment by micaherne** [comment]
Thanks @jlowe64. What list do you mean?

**Comment by jlowe64** [comment]
@micaherne there's a list of about 13 events that I will be adding to the libraries.

**Comment by micaherne** [comment]
@jlowe64 Is the list available somewhere? I'd be interested to know what events are coming up!

**Comment by jlowe64** [comment]
It's currently in a spreadsheet, but we are planning on doing a roadmap for 2.0 soon and it will go into that. Here's the current list of events I'm looking into adding:

```
\mod_forum\event\post_created
\mod_forum\event\subscription_created
\mod_forum\event\subscription_deleted
\mod_forum\event\discussion_subscription_created
\mod_forum\event\discussion_subscription_deleted
\core\event\course_completed
\mod_facetoface\event\signup_success
\assignsubmission_comments\event\comment_created
\assignsubmission_file\event\assessable_uploaded
\assignsubmission_file\event\submission_created
\mod_assign\event\submission_status_viewed
```

The following events are currently being integrated into core by @joshwillcock:

```
\mod_scorm\event\status_submitted
\mod_scorm\event\score_submitted
```

**Comment by ryasmi** [comment]
Closing due to inactivity.

**Comment by micaherne** [comment]
Just in case it's of interest, Moodle 3.6 finally adds this event: \mod_assign\event\course_module_viewed.

https://tracker.moodle.org/browse/MDL-53035

**Comment by ryasmi** [comment]
Interesting thanks @micaherne 👍 I don't have any time to get this added right now, but hopefully someone might pick it up.
